stmhal/onewire.c

```diff
@@ -0,0 +1,825 @@
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+Much of the code was inspired by Derek Yerger's code, though I don't
+think much of that remains.  In any event that was..
+    (copyleft) 2006 by Derek Yerger - Free to distribute freely.
+
+The CRC code was excerpted and inspired by the Dallas Semiconductor
+sample code bearing this copyright.
+//---------------------------------------------------------------------------
+// Copyright (C) 2000 Dallas Semiconductor Corporation, All Rights Reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY,  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL DALLAS SEMICONDUCTOR BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+//
+// Except as contained in this notice, the name of Dallas Semiconductor
+// shall not be used except as stated in the Dallas Semiconductor
+// Branding Policy.
+//--------------------------------------------------------------------------
+*/
+
+#include "onewire.h"
+#include <stdbool.h>
+#include "py/mpconfig.h"
+#include "py/misc.h"
+#include "py/qstr.h"
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/nlr.h"
+#include "pin.h"
+#include "mphal.h"
+#include "std.h"
+
+#define FALSE 0
+#define TRUE  1
+
+typedef struct _onewire_obj_t {
+    mp_obj_base_t   base;
+    const pin_obj_t *pin;
+    GPIO_InitTypeDef initpin;
+    unsigned char ROM_NO[8];
+    uint8_t LastDiscrepancy;
+    uint8_t LastFamilyDiscrepancy;
+    uint8_t LastDeviceFlag;
+} onewire_obj_t;
+
+#define ONEWIRE_INPUT		\
+    self->initpin.Mode = GPIO_MODE_INPUT; \
+    self->initpin.Pull = GPIO_PULLUP; \
+    HAL_GPIO_Init(self->pin->gpio, &(self->initpin))
+
+#define ONEWIRE_OUTPUT      \
+    self->initpin.Mode = GPIO_MODE_OUTPUT_PP; \
+    HAL_GPIO_Init(self->pin->gpio, &(self->initpin))
+    
+#define ONEWIRE_LOW  HAL_GPIO_WritePin(self->pin->gpio, self->pin->pin_mask, 0);
```

Comment:
That's where the nice stuff breaks down. If this module accepts python Pin object, then it should work with it as such - call its .low(), .high(), .input(), .output() methods. This will immediately make it portable to any hardware (and to any user extension, for example, someone may connect GPIO extended over I2C, write Pin class for it, and voila - it all works (modulo timing constraints, that's why they should be configurable too)).